### ISSUE TITLE
how to estimate `max_loops`?

### ISSUE THREAD

**Issue by laimaretto** [ktbyers/netmiko#2368]
Hi!

So, according to [this](https://ktbyers.github.io/netmiko/docs/netmiko/#netmiko.BaseConnection.send_command), the `max_loops` parameter is used inside the `send_command` as a `while` bound before exiting. Inside the `while` loop one can see that:

```
loop_delay=0.2
[...]
while i <= max_loops:
     [...]
     time.sleep(delay_factor * loop_delay)
```
This means that the maximum total time is `max_loops * delay_factor * 0.2`.

Now, let's say that I want to establish a maximum timeout for reading of 400s. Then `max_loops=400/(0.2*delay_factor)`. With `delay_factor=0.5` this leads to `max_loops=4000`.

However I've tried limiting this but even though I've set `max_loops=4000` (or maximum time of 400s) I still can run show commands for over 500s. Why is this happening? 

Finally, I've seen also that `max_loops` is used in `send_config_set()` but there is no `while-loop` inside `send_config_set()`. How does it work in here? Is my estimation above still valid?

Regards!

Lucas

**Comment by ktbyers** [comment]
max_loops and delay_factor are going away in Netmiko 4.x (which is the current `develop` branch). These are being replaced with a `read_timeout` which is an absolute timer for how long `send_command` should keep reading looking for the termination pattern (either the router prompt or the `expect_string` you provide).

So new format will be:

```
net_connect.send_command("show ip bgp", read_timeout=300)
```

Here you would wait for up to 5 minutes (300 seconds) looking for the termination pattern. It will complete as soon as the pattern is detected. If it fails a ReadTimeout exception will be raised.

To answer your specific question, the somewhat easy pattern to use is in Netmiko 3.x

```
Set fast_cli = False as an argument to ConnectHandler
Then set delay_factor=5, that will make the per-loop delay be 1 second:    delay_factor * .2 = 5 * .2 = 1 second
Then set max_loops to number of seconds you want to keep reading for before giving up.
```

With send_config_set() Netmiko loops over sending your commands one at a time. It then tries to read up all the output (which is generally just the configuration commands being echoed). Netmiko expects all of this should be output pretty quickly so there aren't a bunch arguments to directly change this. There is some more nuance related to send_config_set() and how it gathers output, but those are really related to other things (like are you looking for an error_pattern and are you doing cmd_verify).

This should be the math for how to calculate how long it waits before it times out.

https://github.com/ktbyers/netmiko/blob/develop/netmiko/utilities.py#L628

As far as why you can still run commands for over 500s, I would need to see your code example where that is happening (including which device_type you are using). Though you might just want to use `Netmiko4` i.e. the `develop` branch.

**Comment by laimaretto** [comment]
Hi @ktbyers ! Thanks for the reply. So bottom line, I kind of like your suggestion of Netmiko 4.x I'll happily wait for that... :-)

In the mean time, this is a snippet of my code:

```
conn2rtr = ConnectHandler(device_type=deviceType, host=ip, port=port, username=tempUser, password=tempPass, fast_cli = False)
tStart = time.time()
try:
    outRx = conn2rtr.send_config_set(config_commands=inCmd, enter_config_mode=False, cmd_verify=False, delay_factor=delayFactor, max_loops=max_loops)
except Exception as e:
    print(str(e))

tEnd  = time.time()
tDiff = tEnd - tStart
```

As you can see, I mainly use `sen_condfig_set()` with a list of commands. There you can see how I estimate the duration of the run. 

So, for example, for a `max_loops=1000, delay_factor=.5` I would have expected the run to stop at 100s (1000 * 0.2 * 0.5).  It did not finish correctly: There was an exception of `Timed-out reading channel, pattern not found in output: myRouterName` . And it did run for 224s actually.

 I'm using netmiko 3.4.0 with a deivce nokia_sros.

Regards!

Lucas

**Comment by ktbyers** [comment]
Are you classical CLI or model-driven CLI?

What is the value of `inCmd` i.e. what are you passing in here?

`config_commands=inCmd`

**Comment by laimaretto** [comment]
I'm using classical CLI. This is the content of `inCmd`.

```
['# SCRIPT_START', '/environment no more', '/show version', '/show router interface', '/show router bgp routes vpn-ipv4 hunt', 'exit all', '# SCRIPT_FIN']
```

**Comment by ktbyers** [comment]
Okay, so you are trying to use `send_config_set` to send a set of show commands...send_config_set() isn't designed for that and will likely cause you frustrations (which is what you are running into).

Instead of doing that you should probably just use `send_command` and just loop over it:

```python
['/show version', '/show router interface', '/show router bgp routes vpn-ipv4 hunt']
results = {}
for cmd in inCmd:
    # Allow up to 1000 seconds per-command
    output = ssh_conn.send_command(cmd, max_loops=1000, delay_factor=5)
    results[cmd] = output
```

This would create a dictionary where the key was the command and the value was the corresponding output for that command.

**Comment by laimaretto** [comment]
Mmm, I see. Let me try that out. Thanks for the reply!

Now, one question then: what if I need to send both `config`  and/or `show` kind-of commands? Because as per your suggestion, `show` commands should be better off be used with `send_command()` while `config` commands, better off with `send_config_set()`. Is that possible?

Sometimes we build the `inCmd` list with differente kind of commands (`show` and/or `config`, or even `admin`). What would be a better approach for this?

**Comment by ktbyers** [comment]
You would need to separate them into separate operations i.e. send_config_set() for CFG and send_command() for show operations. Admin commands would need to know more about what you were doing.

You could build higher level abstractions in your code if you so chose that allowed you to unify these into one list, but you would need a way to differentiate between cfg and show so that you could dispatch them to Netmiko.

You are going to run into a lot of pain trying to use send_config_set() for show commands or send_command() for a bunch of config operations.

So what I would typically do if I was doing the standard operations of:
1. show commands to check the state of things
2. configuration commands to change things.
3. show commands to verify my config changes

The above combination could be repeated a bunch of times.

I would use `send_command` for all of the show operations and then `send_config_set` for all of the config operations. I would then add logic if I needed to make decisions based on what the show commands told me (rollback the changes, abort the changes, keep going with the changes, et cetera).

Regards, Kirk

**Comment by laimaretto** [comment]
So, I agree with your proposal as a general concept: that's what I usually do, repeat steps 1 to 3 in that order.

I've also followed your suggestion of using send_command() for show commands.

My usual practice is sending, `# START_SCRIPT` at the very beginning, then `all my commands`, and `# END_SCRIPT` at the very last, so I can check if all my commands have been sent over to the router; then store everything (commands and output) in a variable for further analysis. I do this both for `config` and `show` operations.

My original struggle was having incomplete output data ( which I would detect by searching for the strings `# START_SCRIPT` and `# END_SCRIPT` over the received data ): I have solved that pretty much by tunning `max_loops` and `delay_factor`, no matter which method I'm using: either send_command() or send_config_set() would do the job. The thing with send_command() is that I need to do "something extra" to take note of which command produced which output; with `send_config_set()`, I get to see everything automatically.

So I still don't understand totally what is the benefit of using send_command() over send_config_set() for show-type-of commands.

Nevertheless, thanks for this discussion: I've learnt a lot more about Netmiko :-) 

Regards,

Lucas

**Comment by ktbyers** [comment]
But I thought your starting position was...`I can't get send_config_set() to work the way I want it to work`?

**Comment by laimaretto** [comment]
Well, my starting position was `how to estimate max_loops`. 

1. I noticed that I was not getting all the output data from a given run.
2. Tried increasing `delay_factor`: no changes. I've even tried using a `delay_factor=20` in some cases with same result: no complete data.
3. I then discover `max_loops`. Find out that lowering `delay_factor` but increasing `max_loops` solves my issue of very long outputs.
4. That's when I wanted to know how to properly estimate `max_loops` which brings us to the post:
4.1. In the `send_command()` function, I see the `while i <= max_loops` with `time.sleep( 0.2 * delay_factor )`. I don't see this in `send_config_set()` eventhough you can provide it with `max_loops`.
4.2. Additionaly, even if I tried to estimate `max_loops = myTimeOut / ( 0.2 * delay_factor )` I do get the execution to run for more time than `timeout` as I mentioned [here](https://github.com/ktbyers/netmiko/issues/2368#issuecomment-876588956).
5. That's when you said that `send_command()` should be better off used with show commands while `send_config_set()` with config commands.

Bottom line: if I do some tunning of delay_factor and max_loops I can get all I need with `send_config_set()`. Still do not know how to properly estimate `max_loops`. I just know that in order to obtain all my output data, it should be big enough ... :-)

**Comment by ktbyers** [comment]
Fair enough...just don't ask questions of the form `I am using send_config_set() for show commands and it is not working properly`. :-)

This equation does not apply for send_config_set() i.e. that only applies to send_command.

```
max_loops = myTimeOut / ( 0.2 * delay_factor )
```

I am probably going to completely remove max_loops from send_config_set() in Netmiko 4.X. It was used for a different purpose and was probably very rarely used and was being deprecated anyways.

**Comment by ktbyers** [comment]
FYI, Netmiko 4.x also has a new `send_multiline` and `send_multiline_timing` feature where you can send in a list of commands and get the output for each of those commands.

You can see some examples in the test code here:

https://github.com/ktbyers/netmiko/blob/develop/tests/test_netmiko_show.py#L267
